src/testing/snaptest.zig

```diff
@@ -0,0 +1,220 @@
+    ///   - `update_all` const in this file,
+    ///   - `SNAP_UPDATE` env var. 
+    fn should_update(snapshot: *const Snap) bool {
+        return snapshot.update_this or update_all or
+            std.process.hasEnvVarConstant("SNAP_UPDATE");
+    }
+
+    // Compare the snapshot with a formatted string.
+    pub fn diff_fmt(snapshot: *const Snap, comptime fmt: []const u8, fmt_args: anytype) !void {
+        const got = try std.fmt.allocPrint(std.testing.allocator, fmt, fmt_args);
+        defer std.testing.allocator.free(got);
+
+        try snapshot.diff(got);
+    }
+
+    // Compare the snapshot with the json serialization of a `value`.
+    pub fn diff_json(snapshot: *const Snap, value: anytype) !void {
+        var got = std.ArrayList(u8).init(std.testing.allocator);
+        defer got.deinit();
+
+        try std.json.stringify(value, .{}, got.writer());
+        try snapshot.diff(got.items);
+    }
+
+    // Compare the snapshot with a given string.
+    pub fn diff(snapshot: *const Snap, got: []const u8) !void {
+        if (std.mem.eql(u8, got, snapshot.text)) return;
+
+        std.debug.print(
+            \\Snapshot differs.
+            \\Want:
+            \\----
+            \\{s}
+            \\----
+            \\Got:
+            \\----
+            \\{s}
+            \\----
+            \\
+        ,
+            .{
+                snapshot.text,
+                got,
+            },
+        );
+
+        if (!snapshot.should_update()) {
+            return error.SnapDiff;
+        }
+
+        var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
+        defer arena.deinit();
+
+        const allocator = arena.allocator();
+
+        const file_text =
+            try std.fs.cwd().readFileAlloc(allocator, snapshot.location.file, 1024 * 1024);
+        var file_text_updated = try std.ArrayList(u8).initCapacity(allocator, file_text.len);
+
+        const line_zero_based = snapshot.location.line - 1;
+        const range = snap_range(file_text, line_zero_based) orelse
+            return error.SnapNotFound;
+
+        const snapshot_prefix = file_text[0..range.start];
+        const snapshot_text = file_text[range.start..range.end];
+        const snapshot_suffix = file_text[range.end..];
+
+        const indent = get_indent(snapshot_text);
+
+        try file_text_updated.appendSlice(snapshot_prefix);
+        {
+            var lines = std.mem.split(u8, got, "\n");
+            while (lines.next()) |line| {
+                try file_text_updated.writer().print("{s}\\\\{s}\n", .{ indent, line });
+            }
+        }
+        try file_text_updated.appendSlice(snapshot_suffix);
+
+        try std.fs.cwd().writeFile(snapshot.location.file, file_text_updated.items);
+
+        std.debug.print("Updated {s}\n", .{snapshot.location.file});
+        return error.SnapUpdated;
+    }
+};
+
+const Range = struct { start: usize, end: usize };
+fn snap_range(text: []const u8, snap_line: u32) ?Range {
```

Comment:
I'm assuming that string replacement ends up simpler than parsing the AST, mutating, and then rendering it?

Reply:
Exactly: I originally wanted to do `std.zig.Ast` thing here because, wow, I can, but then figured out that Zig's syntax is simple enough to not-overthink it. Added motivational comment.